sympy/physics/quantum/state.py

```diff
@@ -617,6 +619,60 @@ def dual_class(self):
         return TimeDepKet
 
 
+class OrthogonalState(State, StateBase):
+    """General abstract quantum state used as a base class for Ket and Bra."""
+    pass
+
+class OrthogonalKet(OrthogonalState, KetBase):
+    """Orthogonal Ket in quantum mechanics.
+
+    The inner product of two states with different labels will give zero,
+    states with the same label will give one.
+
+        >>> from sympy.physics.quantum import OrthogonalBra, OrthogonalKet, qapply
+        >>> from sympy.abc import m, n
+        >>> (OrthogonalBra(n)*OrthogonalKet(n)).doit()
+        1
+        >>> (OrthogonalBra(n)*OrthogonalKet(n+1)).doit()
+        0
+        >>> (OrthogonalBra(n)*OrthogonalKet(m)).doit()
+        <n|m>
+    """
+
+    @classmethod
+    def dual_class(self):
+        return OrthogonalBra
+
+    def _eval_innerproduct(self, bra, **hints):
+
+        if len(self.args) != len(bra.args):
+            raise ValueError('Cannot multiply a ket that has a different number of labels.')
+
+        for i in range(len(self.args)):
+            diff = self.args[i] - bra.args[i]
+            diff = diff.simplify()
```

Comment:
It is usually better not to call simplify at all because it is an expensive routine. Is this needed for some example in particular?

Reply:
Having complicated ket labels should be an edge case, so in the vast majority of cases `is_zero` should work even without simplification. This code isn't executed on multiplication, it has to be specifically requested via calling `doit`, so the overhead shouldn't be that big.

I decided to use `simplify` because if the `diff` happens to be too hard for `is_zero`, the inner product simply won't resolve and `OrthogonalKet` becomes useless. But if the overhead itronduced by `simplify` is really significant (even for simple expressions which will be the most frequent), we can sacrifice this edge case and just remove `simplify`.

Reply:
Can you give an example where simplify is needed?

I ask because there are many alternatives to simplify. In general simplify is an expensive routine that tries many heuristics and that should be avoided if possible in library code. If we have any idea *what* kind of simplification we want then it is usually better to use that instead.

Reply:
This would work with `expand` instead of `simplify`:
```python
from sympy.abc import a, b
bra = OrthogonalBra( a*(b + 1) )
ket = OrthogonalKet( a*b + a )
(bra*ket).doit()
```
And since I'm not even sure these kets are needed in practice (there are translation operators `|x〉↦ |x+a〉`, but I can't think of an *scaling* operator `|x〉↦ |ax〉`), it seems to me that `expand` will suffice.

Reply:
I went with expand.